modules/process/manager_stacktraces.go

```diff
@@ -0,0 +1,256 @@
+	Stacks   []*Stack   `json:",omitempty"`
+}
+
+// Processes gets the processes in a thread safe manner
+func (pm *Manager) Processes(onlyRoots, noSystem bool, runInLock func()) []*Process {
+	pm.mutex.Lock()
+	processes := make([]*Process, 0, len(pm.processes))
+	if onlyRoots {
+		for _, process := range pm.processes {
+			if noSystem && process.Type == SystemProcessType {
+				continue
+			}
+			if parent, has := pm.processes[process.ParentPID]; !has ||
+				(noSystem && parent.Type == SystemProcessType) {
+				processes = append(processes, process.ToProcess(true))
+			}
+		}
+	} else {
+		for _, process := range pm.processes {
+			if noSystem && process.Type == SystemProcessType {
+				continue
+			}
+			processes = append(processes, process.ToProcess(false))
+		}
+	}
+	if runInLock != nil {
+		runInLock()
+	}
+	pm.mutex.Unlock()
+
+	sort.Slice(processes, func(i, j int) bool {
+		left, right := processes[i], processes[j]
+
+		return left.Start.Before(right.Start)
+	})
+
+	return processes
+}
+
+// ProcessStacktraces gets the processes and stacktraces in a thread safe manner
+func (pm *Manager) ProcessStacktraces(flat, onlyRequests bool) ([]*Process, int64, error) {
+	var stacks *profile.Profile
+	var err error
+	processes := pm.Processes(false, false, func() {
+		reader, writer := io.Pipe()
+		defer reader.Close()
+		go func() {
+			err := pprof.Lookup("goroutine").WriteTo(writer, 0)
+			_ = writer.CloseWithError(err)
+		}()
+		stacks, err = profile.Parse(reader)
+		if err != nil {
+			return
+		}
+	})
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// We cannot use the process pidmaps here because we have released the mutex ...
+	pidMap := map[IDType]*Process{}
+	processStacks := make([]*Process, 0, len(processes))
+	for _, process := range processes {
+		pStack := &Process{
+			PID:         process.PID,
+			ParentPID:   process.ParentPID,
+			Description: process.Description,
+			Start:       process.Start,
+			Type:        process.Type,
+		}
+
+		pidMap[process.PID] = pStack
+		if flat {
+			processStacks = append(processStacks, pStack)
+		} else if parent, ok := pidMap[process.ParentPID]; ok {
+			parent.Children = append(parent.Children, pStack)
+		} else {
+			processStacks = append(processStacks, pStack)
+		}
+	}
+
+	goroutineCount := int64(0)
+
+	// Now walk through the "Sample" slice in the goroutines stack
+	for _, sample := range stacks.Sample {
+		stack := &Stack{}
+
+		// Add the labels
+		for name, value := range sample.Label {
+			if name == DescriptionPProfLabel || name == PIDPProfLabel || (!flat && name == PPIDPProfLabel) || name == ProcessTypePProfLabel {
+				continue
+			}
+			if len(value) != 1 {
+				// Unexpected...
+				return nil, 0, fmt.Errorf("label: %s in goroutine stack with unexpected number of values: %v", name, value)
+			}
+
+			stack.Labels = append(stack.Labels, &Label{Name: name, Value: value[0]})
+		}
+
+		stack.Count = sample.Value[0]
+		goroutineCount += stack.Count
```

Comment:
More general question, what is _general_ value of sample.Value(as well for sample.Label), because the first int seems to be special or so? The documentation from google's profile doesn't seem to mention anything.

Reply:
Both of these are dependent on the profile.  

See:

https://github.com/DataDog/go-profiler-notes/blob/main/pprof.md

https://github.com/google/pprof/blob/master/proto/README.md

https://github.com/google/pprof/blob/master/proto/profile.proto#L108-L117